compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala

```diff
@@ -234,8 +243,11 @@ object ProtoTypes {
   extends UncachedGroundType with ApplyingProto with FunOrPolyProto {
     override def resultType(implicit ctx: Context): Type = resType
 
-    def isMatchedBy(tp: Type)(implicit ctx: Context): Boolean =
-      typer.isApplicable(tp, Nil, unforcedTypedArgs, resultType)
+    def isMatchedBy(tp: Type, keepConstraint: Boolean)(implicit ctx: Context): Boolean = {
+      val args = unforcedTypedArgs
+      def isPoly(tree: Tree) = tree.tpe.widenSingleton.isInstanceOf[PolyType]
+      typer.isApplicable(tp, Nil, args, resultType, keepConstraint && !args.exists(isPoly))
```

Comment:
Why the special cases when one of the argument is polymorphic ?

Reply:
It's in the comment.